torch/lib/c10d/logger.cpp

```diff
@@ -85,7 +91,145 @@ void Logger::set_construction_data_and_log(
   LogPyTorchDDPUsage(*ddp_logging_data_);
 }
 
+void Logger::calculate_avg_cpu_time(
+    int64_t& avg_time,
+    int64_t cpu_start_time,
+    int64_t cpu_end_time) {
+  // If cpu_end_time is not recorded in this iteration,
+  // avg_time will return invalid value.
+  // For some cases like DDP runs on non-sync mode, backward compute
+  // end time can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  if (cpu_end_time < cpu_start_time) {
+    return;
+  }
+  long num_iters = reducer_->num_iterations_;
+  avg_time = ((cpu_end_time - cpu_start_time) + avg_time * (num_iters - 1)) /
+      num_iters;
+}
+
+#ifdef USE_CUDA
+void Logger::calculate_avg_gpu_time(
+    int64_t& avg_time,
+    at::cuda::CUDAEvent& gpu_start,
+    at::cuda::CUDAEvent& gpu_end) {
+  TORCH_CHECK(reducer_->num_iterations_ > 0);
+  long num_iters = reducer_->num_iterations_;
+  float milliseconds = gpu_start.elapsed_time(gpu_end);
+  // If gpu_end is not recorded in this iteration,
+  // milliseconds will have invalid value.
+  // For some cases like DDP runs on non-sync mode,
+  // gpu_end can not be recorded in this iteration and thus can not
+  // calculate the valid avg_time.
+  // In this case, skip calculating the avg_time and return.
+  if (milliseconds < 0) {
+    return;
+  }
+  avg_time = (int(milliseconds * kMilliSecondToNanosSecond) +
+              avg_time * (num_iters - 1)) /
+      num_iters;
+}
+#endif
+
+void Logger::set_runtime_stats() {
+  // Sync with reducer's data
+  std::lock_guard<std::mutex> lock(reducer_->mutex_);
+  // set runtime stats after 1st iteration is complete.
+  if (reducer_->num_iterations_ == 0) {
+    return;
+  }
+  // Set ith iteration when the runtime stats are set.
+  ddp_logging_data_->iteration = reducer_->num_iterations_;
+  // If unused_parameters_ is not empty, calculate its sizes.
+  // unused_parameters_ is calculated in forward call of
+  // each iteration.
+  for (const auto& unused_index : reducer_->unused_parameters_) {
```

Comment:
This might trigger a nontrivial perf hit if significant (1M+ for example) parts of the model (maybe in something like multi-headed attention) aren't used in certain iterations.

Reply:
be noted here the unused_parameters_ are parameter tensors, the number of parameter tensors is usually small. e.g., for 1G bert model, totally it has 150 parameter tensors.